src/utils.c

```diff
@@ -110,30 +117,208 @@ SEXP colnamesInt(SEXP x, SEXP cols, SEXP check_dups) {
   UNPROTECT(protecti);
   return ricols;
 }
 
+/* replace_dot_alias
+ * we don't just simply alias .=list because i) list is a primitive (faster to iterate) and ii) we test for use
+ * of "list" in several places so it saves having to remember to write "." || "list" in those places
+ * replace_dot_aliasR wrapper should be used to prevent update of input
+ */
+SEXP replace_dot_alias(SEXP x) {
+  if (isLanguage(x) && !isFunction(CAR(x))) {
+    if (CAR(x) == sym_bquote) // handling `.` inside bquote, #1912
+      return x;
+    if (CAR(x) == sym_dot)
+      SETCAR(x, sym_list);
+    for (SEXP t=x; t!=R_NilValue; t=CDR(t)) {
+      if (!isNull(CADR(t)))
+        SETCADR(t, replace_dot_alias(CADR(t)));
+    }
+  }
+  return x;
+}
+SEXP replace_dot_aliasR(SEXP x) {
+  return replace_dot_alias(MAYBE_REFERENCED(x) ? duplicate(x) : x);
+}
+
+/* range_int
+ * used in exprCols(v4:v2) to produce 4:2 sequence
+ */
+SEXP range_int(int from, int to) {
+  if (from==NA_INTEGER || to==NA_INTEGER)
+    error(_("internal error: NA was supplied to generate integer range")); // # nocov
+  int len = to - from;
+  int ord = len>=0 ? 1 : -1;
+  len = len * ord;
+  SEXP ans = PROTECT(allocVector(INTSXP, len+1));
+  int *ansp = INTEGER(ans);
+  for (int i=0; i<=len; i++)
+    ansp[i] = from+i*ord;
+  UNPROTECT(1);
+  return ans;
+}
+
+/* funCols
+ * helper that calls lapply(DT, fun) to produce logical vector
+ */
+SEXP funCols(SEXP x, SEXP fun, SEXP rho) {
+  SEXP lfn = PROTECT(LCONS(sym_lapply, LCONS(x, LCONS(fun, R_NilValue))));
+  SEXP lst = PROTECT(eval(lfn, rho));
+  SEXP log = PROTECT(allocVector(LGLSXP, length(x)));
+  int *logp = LOGICAL(log);
+  for (int i=0; i<length(lst); ++i) {
+    SEXP lsti = VECTOR_ELT(lst, i);
+    if (length(lsti)!=1 || !isLogical(lsti) || LOGICAL(lsti)[0]==NA_LOGICAL)
+      error("When argument for columns selection is a function, it is applied to each column; the output of this function must be a non-missing boolean scalar signalling inclusion/exclusion of the column. However, these conditions were not met at least for: %s", CHAR(STRING_ELT(getAttrib(x, R_NamesSymbol), i)));
+    logp[i] = LOGICAL(lsti)[0];
+  }
+  UNPROTECT(3);
+  return log;
+}
+
+/* exprCols
+ * turns NSE expression into columns int
+ * helper for j and .SDcols
+ * handles v1, "v1", v1:v3, !"v1", -"v1", paste0("v",1:3)
+ * does not evaluate only for `symbol:symbol` expression, or `symbol` when "symbol" is existing column
+ */
+SEXP exprCols(SEXP x, SEXP expr, /*SEXP with, */SEXP rho) {
+  int protecti=0;
+  if (isNull(expr))
+    error(_("columns selection is NULL")); // expr=NULL
+  SEXP sym_parenthesis = install("(");
+  SEXP sym_patterns = install("patterns");
+  while(isLanguage(expr) && CAR(expr)==sym_parenthesis)
+    expr = CADR(expr);
+  if (isNull(expr))
+    error(_("columns selection is NULL")); // expr=((NULL))
+  // TODO handle inverse selection: !cols_var, !"V2", !V2, !paste0("V",2:3)
```

Comment:
```.SDcols``` currently accepts ```!V3:V2```. The backwards ordering has been added to table.

Also, in R, I think this is cleaner:
```
negate_cols = is.call(colsub) && deparse(colsub[[1L]], 500L, backtick=FALSE) %chin% c("!", "-")
if (negate_cols) colsub = colsub[[2L]]
```
Not sure if this is valid syntax in C:
```
bool inverse = isLanguage(expr) && (CAR(expr)==sym_bang || CAR(expr)==sym_minus);
if (inverse) expr = CADR(expr);

Reply:
The concern is not ordering !V3:V2 vs !V2:V3 but complexity of handling `:` and `!`/`-`. Colon `:` is handled specially, and `!` and `-` are handled specially too. There can be multiple combinations of both of them that would need to be addressed if we would allow that. From `!((V2:V3))`, `-{({(V2:V3)})}`, `!V2:3`, `!V2:"V3"`, etc. We have to draw a line somewhere.

Why you think it is cleaner? I am actually not really convinced it to be cleaner.

Reply:
Sorry - I don't think the R way is cleaner than C. The cleaner part is initializing ```inverse``` as the ```isLanguage(expr) && ...``` part.

Reply:
Oh, I got it now. I misread your previous comment. Yes it looks better! will try it. Thanks

Reply:
Current tests should be a good reference of what to support, for example !(V3:V2) already works but !V3:V2 does not

Reply:
```!V3:V2``` works in ```.SDcols``` but not in ```j```. While there is no specific test that is ```.SDcols = !Var_High:Var_low```. I think it would be better to keep by, j, and .SDcols aligned except where it doesn't make sense which would be certain calls.

For example, ```c()``` can be dangerous in ```j``` but there should not be any problems in ```by``` or ```.SDcols```:
```
dt = as.data.table(lapply(1:5, c))
dt[, c(V1 + V2, lapply(.SD, '*', 2))]

      V1    V1    V2    V3    V4    V5
   <int> <num> <num> <num> <num> <num>
1:     3     2     4     6     8    10
```
I see the mode telling ```CexprCols``` whether to evaluate the expression at all. 

1. .SDcols always gets evaluated. 
2. by gets evaluated on calls being: ```c("c","key","names", "intersect", "setdiff"))```, a call of ```:``` without ```.N``` or ```nrow()```, or if the expression is a symbol. 
3. j gets evaluated similar to by but ```c()``` may cause problems.

And the ```[data.table``` function would have something like:
```
orig_jsub = jsub
jsub = exprCols(...)
jsub_needs_eval = orig_jsub == jsub
```

I will add some calls to the giant table above.